### DISCUSSION TITLE
Simplify def variances()?

### DISCUSSION THREAD

**Discussion by HDembinski** [scikit-hep/boost-histogram#701] [General]
```py
        if hasattr(view, "sum_of_weights"):
            valid = view.sum_of_weights**2 > view.sum_of_weights_squared  # type: ignore[union-attr]
            return np.divide(  # type: ignore[no-any-return]
                view.variance,  # type: ignore[union-attr]
                view.sum_of_weights,  # type: ignore[union-attr]
                out=np.full(view.sum_of_weights.shape, np.nan),  # type: ignore[union-attr]
                where=valid,
            )

        if hasattr(view, "count"):
            return np.divide(  # type: ignore[no-any-return]
                view.variance,  # type: ignore[union-attr]
                view.count,  # type: ignore[union-attr]
                out=np.full(view.count.shape, np.nan),  # type: ignore[union-attr]
                where=view.count > 1,  # type: ignore[union-attr]
            )

        return view.variance  # type: ignore[union-attr]
```
Why is this written with np.divide? Can't this be simplified?

In the second part, `view.count > 0` seems more obvious. `view.count > 1` implies that you know how view.variance is implemented (which may in fact return NaN if view.count == 1).

**Comment by henryiii** [comment]
We had a long email thread about this - `np.divide` with a mask is fast. It's also quite simple, really, just a single function call with four arguments; the type ignores are due to the fact that we don't have this Generic, so MyPy can't tell what kind of view we have - we'd be stuck with them in any other form too. Though I might be able to teach MyPy about the if statements by using a Protocol, come to think of it.

It also gives us the freedom to control the exact form of the result. In this case, if the count is exactly 1, we declare that we don't know the variance and return NaN, regardless of whether `view.variance` is NaN or not - this is much less error prone for users who do not read the spec and notice they are supposed to gate all calculations on the value of counts (as you saw in the recent Gitter thread, no one was reading the docs) if view.variance is not NaN, and also is consistent. And we also get to save a little time by not bothering to calculate values that are not needed.
